### PR TITLE
Bank transaction history

### PR DESCRIPTION
**Pull request by martinmilev**
[presentation](https://martinmilev.github.io/htmlPages/servlets/bankTrHr/bank.html)
Ready for review. Tests uncompleted

### CODE REVIEW

**Comment by freekman** [comment]
i like the design but remove the logout from here

File: src/main/java/com/clouway/core/AccountRepository.java [hunk 1]
```diff
@@ -8,6 +8,8 @@
 public interface AccountRepository {
   void register(Account account);
 
+  void updateAmount(Double amount, Account account);
```
Comment (freekman) [inline src/main/java/com/clouway/core/AccountRepository.java]: .

File: src/main/java/com/clouway/core/HistoryRepository.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.clouway.core;
+
+import java.util.List;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface HistoryRepository {
+  List<List<Transaction>> getPages(String username);
```
Comment (freekman) [inline src/main/java/com/clouway/core/HistoryRepository.java]: please soot me ```List<List ..... >```

Comment (freekman) [inline src/main/java/com/clouway/core/HistoryRepository.java]: what was wrong with List<Transaction> getPage(Integer limit, Integer offset)

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentHistoryRepository.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import com.clouway.persistent.datastore.DataStore;
+import com.google.common.collect.Lists;
+
+import java.sql.SQLException;
+import java.util.List;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class PersistentHistoryRepository implements HistoryRepository {
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentHistoryRepository.java]: tell me if you are not writing TDD now .When  will you do it.
MISSING TESTS

File: src/main/java/com/clouway/core/Transaction.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package com.clouway.core;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class Transaction {
+  public String date;
```
Comment (freekman) [inline src/main/java/com/clouway/core/Transaction.java]: FINAL

File: src/main/java/com/clouway/http/servlets/LoginPageServlet.java [hunk 5]
```diff
@@ -59,11 +60,14 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
     String name = req.getParameter("name");
     String pswd = req.getParameter("password");
+
     Optional<Account> optional = repository.getByName(name);
 
     if (optional.isPresent()) {
       Account account = optional.get();
       if(pswd.equals(account.password)) {
+
+        req.getSession().setAttribute("name", name);
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/LoginPageServlet.java]: .

File: src/main/resources/account.html [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE html>
```
Comment (freekman) [inline src/main/resources/account.html]: not needed in this PR

File: src/main/resources/history.html [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            <div class="btn-group" role="group">
+                <button type="submit" name="button" value="prev" type="submit" class="btn btn-info">Prev
+                </button>
+                <button type="submit" name="button" value="next" type="submit" class="btn btn-info">Next
+                </button>
+                <p class="btn btn-default active">Page ${current}/${pageNum}</p>
+                <p class="btn btn-default active">Username: ${name}</p>
+                <p class="btn btn-default active">Balance: ${balance}</p>
+                <button type="submit" name="button" value="trans" type="submit" class="btn btn-primary">Back</button>
+                <button type="submit" name="button" value="logout" type="submit" class="btn btn-danger">Logout
```
Comment (freekman) [inline src/main/resources/history.html]: remove the logout.For the Back use link

File: src/main/resources/history.html [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+<head>
+    <meta charset="UTF-8">
+    <title>Transaction History</title>
+    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
+</head>
+<body>
+<div class="container-fluid text-center">
+    <h1>Transaction History</h1>
+    <div>
+        <form method="post" action="/history">
```
Comment (freekman) [inline src/main/resources/history.html]: why is this post are you sure it is a correct REST design

File: src/main/resources/history.html [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                <button type="submit" name="button" value="trans" type="submit" class="btn btn-primary">Back</button>
+                <button type="submit" name="button" value="logout" type="submit" class="btn btn-danger">Logout
+                </button>
+            </div>
+        </form>
+    </div>
+</div>
+<div class="container-fluid col-md-8 col-md-offset-2">
+    <table class="table table-bordered">
+        <thead>
```
Comment (freekman) [inline src/main/resources/history.html]: this is really ugly one solution here could be to add```${history}```

And add some abstraction like the HTML Template .

File: src/main/java/com/clouway/core/HistoryRepository.java (hunk 2 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/HistoryRepository.java]: easier to iterate

**Comment by freekman** [comment]
is this PR ready for review

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentHistoryRepository.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  private final DataStore dataStore;
+
+  public PersistentHistoryRepository(DataStore dataStore) {
+    this.dataStore = dataStore;
+  }
+
+  public Integer getNumOfRows(String username) {
+    String query = "select count(1) from transaction_history where Name='" + username + "'";
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentHistoryRepository.java]: explaine to what does this query  do ```select count(1) from```` and what is this ```1```

File: src/main/java/com/clouway/core/HistoryRepository.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.List;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface HistoryRepository {
+  Integer getNumOfRows(String username);
+
+  List<Transaction> getTransactionsForAccount(String username, Integer limit, Integer offset);
```
Comment (freekman) [inline src/main/java/com/clouway/core/HistoryRepository.java]: bad naming

File: src/main/test/com/clouway/core/HtmlTableBuilderTest.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class HtmlTableBuilderTest {
+  @Test
+  public void happyPath() throws Exception {
+    HtmlTableBuilder builder = new HtmlTableBuilder();
+    Date date = new Date();
+    List<Transaction> transaction = Lists.newArrayList(new Transaction(" ", "John", "Deposit", 10D));
```
Comment (freekman) [inline src/main/test/com/clouway/core/HtmlTableBuilderTest.java]: add The Transactuion date

File: src/main/resources/history.html [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        <tr>
+            <th>Date</th>
+            <th>Operation</th>
+            <th>Amount</th>
+        </tr>
+        </thead>
+        <tbody>
+            ${rows}
+        <!--<tr>
```
Comment (freekman) [inline src/main/resources/history.html]: remove commented code

File: src/main/java/com/clouway/http/servlets/AccountPageServlet.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class AccountPageServlet extends HttpServlet {
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/AccountPageServlet.java]: why am i still seeing this the PR was tagged need's review this is not removed .

File: src/main/java/com/clouway/core/HistoryRepository.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package com.clouway.core;
+
+import java.util.List;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface HistoryRepository {
+  Integer getNumOfRows(String username);
```
Comment (martinmilev) [inline src/main/java/com/clouway/core/HistoryRepository.java]: I use it to determinate the last page and not show empty pages.

**Comment by freekman** [comment]
I will not review this PR until the old comments are not fixed!

File: src/main/java/com/clouway/persistent/adapter/jdbc/PersistentHistoryRepository.java (hunk 10 above)
Comment (martinmilev) [inline src/main/java/com/clouway/persistent/adapter/jdbc/PersistentHistoryRepository.java]: It's a faster way to count the rows. Select(1) column of a row with value "1" for each row in the table, so it doesn't read the actual values in the table.

File: src/main/java/com/clouway/core/HistoryRepository.java (hunk 11 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/HistoryRepository.java]: `List<Transaction> get(...);` 
?

File: src/main/java/com/clouway/core/HtmlTableBuilder.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class HtmlTableBuilder implements TableBuilder {
+  @Override
+  public String getRows(List<Transaction> transactions) {
+    StringBuilder html = new StringBuilder();
+    for (Transaction tr : transactions) {
+      html.append("<tr><td>" + tr.date + "</td><td>" + tr.operation + "</td><td>" + tr.amount + "</td>");
```
Comment (martinmilev) [inline src/main/java/com/clouway/core/HtmlTableBuilder.java]: Missed a closing tag. 
fixed it.

File: src/main/java/com/clouway/core/TableBuilder.java [hunk 17]
```diff
@@ -0,0 +1,10 @@
+package com.clouway.core;
+
+import java.util.List;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface TableBuilder {
```
Comment (martinmilev) [inline src/main/java/com/clouway/core/TableBuilder.java]: fixed.

File: src/main/resources/history.html (hunk 13 above)
Comment (martinmilev) [inline src/main/resources/history.html]: fixed.

File: src/main/java/com/clouway/core/AccountRepository.java [hunk 18]
```diff
@@ -8,6 +8,7 @@
 public interface AccountRepository {
   void register(Account account);
 
-  Optional<Account> getByName(String name);
-}
+  void updateAmount(Double amount, String name);
```
Comment (martinmilev) [inline src/main/java/com/clouway/core/AccountRepository.java]: fixed.

**Pull request closed without merging**